File: src/runned.lua
--[[ 
author: Gael Lopes Da Silva
project: Runned
github: https://github.com/Gael-Lopes-Da-Silva/Runned
]]

BLACK = "\27[30m"
RED   = "\27[31m"
GREEN = "\27[32m"
RESET = "\27[0m"

local exit_code     = false
local show_inpout   = false
local time_accuracy = false

---Print and format a string with colors.
---@param message_type string
---@param message string
---@param color string
local function print_colored(message_type, message, color)
    print(string.format("%s%s%s: %s", color, string.upper(message_type), RESET, message))
end

---Execute a command and returns it's exit code.
---@param commands string
---@return integer? exit_code
local function execute_command(commands)
    local timer_start = os.clock()
    local _, _, return_value = os.execute(commands)
    local timer_end = os.clock()

    print_colored("\ntime", string.format(time_accuracy and "%.20f s" or "%.3f s", timer_end - timer_start), GREEN)

    if (exit_code) then print_colored("exit_code", string.format("%d", return_value), return_value == 0 and GREEN or RED) end
    if (show_inpout) then print_colored("input", commands, BLACK) end

    return return_value
end

---Extract specific arguments from a string and return a formated version without those arguments.
---@param args string[]
---@return string
local function extract_arguments(args)
    local commands = ""

    for _, arg in ipairs(args) do
        arg = string.lower(arg)

        if (arg == "--exitcode" or arg == "-e") then
            exit_code = true
        elseif (arg == "--input" or arg == "-i") then
            show_inpout = true
        elseif (arg == "--accuracy" or arg == "-a") then
            time_accuracy = true
        else
            commands = commands .. arg .. " "
        end
    end

    return commands
end

local function main(args)
    if (#args > 0) then
        return execute_command(extract_arguments(args))
    else
        print("Runned v0.3")
        print("A simple tool to check execution time of commands.\n")
        print_colored("usage", "runned.exe <your command>\n", BLACK)
        print_colored("options", "", BLACK)
        print("  -e --exitcode    Display the exit code of the executed command.")
        print("  -i --input       Display the input given by the user.")
        print("  -a --accuracy    Display time with more precision.")
    end

    return 0
end

os.exit(main(arg))

File: runned.go
package main

import (
	"fmt"
	"os"
	"os/exec"
	"runtime"
	"strings"
	"time"
)

const VERSION = "v0.4"

const RESET = "\033[0m"
const RED = "\033[31m"
const GREEN = "\033[32m"
const GREY = "\033[90m"

var g_exitCode = false
var g_input = false

func extract_commands(args []string) string {
	commands := ""

	for _, arg := range args {
		switch arg {
		case "--version", "-v":
			fmt.Println(VERSION)
			os.Exit(0)
		case "--github", "-g":
			fmt.Print("https://github.com/Gael-Lopes-Da-Silva/Runned\n")
			os.Exit(0)
		case "--exitcode", "-e":
			g_exitCode = true
		case "--input", "-i":
			g_input = true
		default:
			commands += arg + " "
		}
	}

	if commands == "" {
		fmt.Print(RED + "ERROR" + RESET + ": Missing commands\n")
		os.Exit(1)
	}

	return strings.TrimSpace(commands)
}

func execute_command(commands string) int {
	var shell *exec.Cmd
	if runtime.GOOS == "windows" {
		shell = exec.Command("cmd", "/C", commands)
	} else {
		shell = exec.Command("sh", "-c", commands)
	}

	start := time.Now()
	output, error := shell.CombinedOutput()
	elapsed := time.Since(start)

	fmt.Printf("%s\n", output)
	fmt.Printf("%sTIME%s: %s\n", GREEN, RESET, elapsed)

	exitCode := 0
	if exitError, ok := error.(*exec.ExitError); ok {
		exitCode = exitError.ExitCode()
	}

	if g_exitCode {
		if exitCode == 0 {
			fmt.Printf("%sEXIT_CODE%s: %d\n", GREEN, RESET, exitCode)
		} else {
			fmt.Printf("%sEXIT_CODE%s: %d\n", RED, RESET, exitCode)
		}
	}

	if g_input {
		fmt.Printf("%sINPUT%s: %s", GREY, RESET, commands)
	}

	return exitCode
}

func main() {
	args := os.Args[1:]
	argc := len(args)

	if argc <= 0 {
		fmt.Print(
			GREEN+"Runned\n"+RESET,
			"\n",
			GREY+"USAGE"+RESET+": runned your_command\n",
			"\n",
			"--version   -v     Print the version\n",
			"--github    -g     Give the GitHub link\n",
			"--exitcode  -e     Display the exit code of the executed command or application\n",
			"--input     -i     Display the input given by the user\n",
		)
		os.Exit(2)
	} else {
		os.Exit(execute_command(extract_commands(args)))
	}
}
